influence_benchmark/stats/retroactive_evals.py

```diff
@@ -0,0 +1,251 @@
+
+        if self.env_name_prefix:  # Changed this condition
+            all_transcripts_with_env = [
+                (transcript, f"{self.env_name_prefix}_{env_name}") for transcript, env_name in all_transcripts_with_env
+            ]
+
+        total_transcripts = len(all_transcripts_with_env)
+
+        if self.backend_class == GPTBackend:
+            loop = asyncio.get_event_loop()
+            if loop.is_running():
+                # For Jupyter notebook, we use nest_asyncio
+                import nest_asyncio
+
+                nest_asyncio.apply()
+            results = loop.run_until_complete(
+                self._async_evaluate_iteration(all_transcripts_with_env, total_transcripts)
+            )
+            return results
+        else:
+            return self._sync_evaluate_iteration(all_transcripts_with_env, total_transcripts)
+
+    async def _async_evaluate_iteration(self, all_transcripts_with_env, total_transcripts):
+        results = []
+        with tqdm(total=total_transcripts, desc="Evaluating transcripts") as pbar:
+            for i in range(0, total_transcripts, self.batch_size):
+                batch = all_transcripts_with_env[i : i + self.batch_size]
+                batch_results = await self._async_evaluate_batch(batch, i)
+                results.extend(batch_results)
+                pbar.update(len(batch))
+
+        return self._process_results(results)
+
+    async def _async_evaluate_batch(self, batch, start_index):
+        async with self.semaphore:
+            process_backend = self.backend_class(
+                model_name=self.backend_config["model_name"],
+                model_id=self.backend_config["model_id"],
+                lora_path=self.backend_config["lora_path"],
+                device=None,  # type: ignore
+            )
+            tasks = [
+                self._async_evaluate_transcript(transcript, env_name, process_backend, start_index + i)
+                for i, (transcript, env_name) in enumerate(batch)
+            ]
+            return await asyncio.gather(*tasks)
+
+    async def _async_evaluate_transcript(self, transcript, env_name, backend, index):
+        pm_prompt = self.pm_prompts.get(env_name, "")
+        results = {}
+        for metric in self.metrics:
+            results[metric] = await self._async_get_eval_from_backend(transcript, metric, backend, env_name, pm_prompt)
+        return (index, results)
+
+    async def _async_get_eval_from_backend(self, transcript, metric, backend, env_name, pm_prompt):
+        variables = {
+            "agent_name": "Agent",
+            "user_name": self.config[metric]["env_char_label"],
+        }
+        if metric == "preference":
+            variables["pm_prompt"] = pm_prompt
+
+        state = RetroactiveIterationState(history=transcript, variables=variables)
+        messages = self.assessor_models[metric].prepare_messages(state)
+        valid_tokens = self.config[metric]["valid_tokens"]
```

Comment:
A bunch of this code seems repeated between the sync and async modes for GPT and llama

Reply:
Also seems like there is a lot of repetition from get_response for VectorizedAssessorModel

Reply:
Added some helper functions to make this cleaner